File: planners/cpp/cppplanner_upf.cxx
#include "cppplanner_upf.hxx"

#include "planner.hxx"


std::optional<std::vector<std::string>> solve(upf::Problem& problem, std::function<double(std::set<std::string>)> heuristic)
{
  std::vector<cppplanner::Action> actions;
  for (auto& a : problem.actions()) {
    actions.emplace_back(a.name(), a.precondition(), a.positive_effect(), a.negative_effect());
  }
  cppplanner::Problem planner_problem(actions, problem.init(), problem.goal());
  return cppplanner::solve(planner_problem, heuristic);
}

File: planners/cpp/cppplanner_upf.py
import _cppplanner_pyupf as cpp

def _upf_action_to_cpp(action):
    return cpp.Action(action.name, action.precondition, action.positive_effect, action.negative_effect)

def _upf_problem_to_cpp(problem):
    actions = [_upf_action_to_cpp(x) for x in problem.actions]
    return cpp.Problem(actions, problem.init, problem.goal)

def solve(problem):
    return cpp.solve(_upf_problem_to_cpp(problem))

def solve_with_heuristic(problem, h):
    return cpp.solve_with_heuristic(_upf_problem_to_cpp(problem), h)
